File: C++/6-LinkedList/Problems/2-MergeTwoSortedLists/MergeTwoSortedListsPro.js
// Definition of a ListNode (used for all approaches)
function ListNode(val, next) {
    this.val = (val===undefined ? 0 : val)
    this.next = (next===undefined ? null : next)
}

// Helper function to create a linked list from an array (used for testing)
function createLinkedList(arr) {
    if (!arr || arr.length === 0) return null;
    let head = new ListNode(arr[0]);
    let current = head;
    for (let i = 1; i < arr.length; i++) {
        current.next = new ListNode(arr[i]);
        current = current.next;
    }
    return head;
}

// Helper function to convert a linked list to an array (used for output and verification)
function linkedListToArray(head) {
    const result = [];
    let current = head;
    while (current) {
        result.push(current.val);
        current = current.next;
    }
    return result;
}

// Helper function to visualize linked list
function printLinkedList(head) {
  let current = head;
  let str = "";
  while (current) {
    str += current.val + " -> ";
    current = current.next;
  }
  str += "null";
  console.log(str);
}

/**
 * Merge two sorted linked lists.
 *
 * Input:
 * - l1: The head of the first sorted linked list.
 * - l2: The head of the second sorted linked list.
 *
 * Output:
 * - The head of the merged sorted linked list.
 */

// Approach 1: Iterative with Dummy Node
// Time Complexity: O(m + n), where m and n are the lengths of the two lists.
// Space Complexity: O(1) - Constant space.
const mergeTwoListsV1 = function(l1, l2) {
    let dummy = new ListNode(-1); // Dummy node to simplify the head handling
    let tail = dummy;

    while (l1 && l2) {
        if (l1.val <= l2.val) {
            tail.next = l1;
            l1 = l1.next;
        } else {
            tail.next = l2;
            l2 = l2.next;
        }
        tail = tail.next;
    }

    // Append the remaining non-null list
    tail.next = l1 || l2;

    return dummy.next; // Return the merged list starting from the node after the dummy
};

// Approach 2: Recursive
// Time Complexity: O(m + n)
// Space Complexity: O(m + n) - Due to the recursive call stack.
const mergeTwoListsV2 = function(l1, l2) {
    if (!l1) return l2;
    if (!l2) return l1;

    if (l1.val <= l2.val) {
        l1.next = mergeTwoListsV2(l1.next, l2);
        return l1;
    } else {
        l2.next = mergeTwoListsV2(l1, l2.next);
        return l2;
    }
};

// Approach 3: Iterative In-Place (No Dummy Node)
// Time Complexity: O(m + n)
// Space Complexity: O(1) - Constant space.  Modifies the original lists.
const mergeTwoListsV3 = function(l1, l2) {
    if (!l1) return l2;
    if (!l2) return l1;

    // Determine the head of the merged list
    if (l1.val > l2.val) {
        [l1, l2] = [l2, l1]; // Swap l1 and l2
    }
    let head = l1; // head of the merged list
    let prev = l1;
    l1 = l1.next;

    while (l1 && l2) {
        if (l1.val <= l2.val) {
            prev.next = l1;
            l1 = l1.next;
        } else {
            prev.next = l2;
            let temp = l2.next;
            l2.next = l1;
            l2 = temp;
        }
        prev = prev.next;
    }
    prev.next = l1 || l2; // append what is left of l1 or l2
    return head;
};

// Approach 4: Using Array and Sorting (Not a true linked list merge, but for comparison)
// Time Complexity: O((m + n) log (m + n)) - Due to sorting.
// Space Complexity: O(m + n) - To store the combined array.
const mergeTwoListsV4 = function(l1, l2) {
    const arr1 = linkedListToArray(l1);
    const arr2 = linkedListToArray(l2);
    const mergedArray = arr1.concat(arr2);
    mergedArray.sort((a, b) => a - b); // Sort the merged array
    return createLinkedList(mergedArray);
};

// Approach 5: Iterative with Carry (For demonstration -  not directly applicable here, but shows a different pattern)
// This approach is more relevant for merging sorted arrays where you might have
// a carry-over, like in adding two numbers represented as linked lists.  It's
// included here to show a different linked list manipulation pattern.  It does NOT correctly merge sorted lists.
// Time Complexity: O(m+n)
// Space Complexity: O(1)
const mergeTwoListsV5 = function(l1, l2) {
    let dummy = new ListNode(-1);
    let current = dummy;
    let carry = 0;

    while (l1 || l2 || carry) {
        const val1 = l1 ? l1.val : 0;
        const val2 = l2 ? l2.val : 0;
        const sum = val1 + val2 + carry;

        carry = Math.floor(sum / 10);  // Corrected carry calculation.
        current.next = new ListNode(sum % 10); // Corrected node creation.
        current = current.next;

        l1 = l1 ? l1.next : null;
        l2 = l2 ? l2.next : null;
    }
    return dummy.next;
};



// Test Cases
const list1 = createLinkedList([1, 2, 4]);
const list2 = createLinkedList([1, 3, 4]);
const list3 = createLinkedList([]);
const list4 = createLinkedList([]);
const list5 = createLinkedList([5]);
const list6 = createLinkedList([1,2,3,4,5,6]);

console.log("Input Lists:");
printLinkedList(list1);
printLinkedList(list2);
console.log("\n");

// Test Case 1: Basic merge
console.log("Test Case 1: Basic Merge");
let merged1_1 = mergeTwoListsV1(createLinkedList([1, 2, 4]), createLinkedList([1, 3, 4]));
let merged1_2 = mergeTwoListsV2(createLinkedList([1, 2, 4]), createLinkedList([1, 3, 4]));
let merged1_3 = mergeTwoListsV3(createLinkedList([1, 2, 4]), createLinkedList([1, 3, 4]));
let merged1_4 = mergeTwoListsV4(createLinkedList([1, 2, 4]), createLinkedList([1, 3, 4]));
let merged1_5 = mergeTwoListsV5(createLinkedList([1, 2, 4]), createLinkedList([1, 3, 4]));
console.log("V1:", linkedListToArray(merged1_1)); // Expected: [1, 1, 2, 3, 4, 4]
console.log("V2:", linkedListToArray(merged1_2)); // Expected: [1, 1, 2, 3, 4, 4]
console.log("V3:", linkedListToArray(merged1_3)); // Expected: [1, 1, 2, 3, 4, 4]
console.log("V4:", linkedListToArray(merged1_4)); // Expected: [1, 1, 2, 3, 4, 4]
console.log("V5:", linkedListToArray(merged1_5)); // Note: This is NOT a correct merge.
console.log("\n");

// Test Case 2: One list is empty
console.log("Test Case 2: One List Empty");
let merged2_1 = mergeTwoListsV1(createLinkedList([1, 2, 4]), createLinkedList([]));
let merged2_2 = mergeTwoListsV2(createLinkedList([1, 2, 4]), createLinkedList([]));
let merged2_3 = mergeTwoListsV3(createLinkedList([1, 2, 4]), createLinkedList([]));
let merged2_4 = mergeTwoListsV4(createLinkedList([1, 2, 4]), createLinkedList([]));
let merged2_5 = mergeTwoListsV5(createLinkedList([1, 2, 4]), createLinkedList([]));
console.log("V1:", linkedListToArray(merged2_1)); // Expected: [1, 2, 4]
console.log("V2:", linkedListToArray(merged2_2)); // Expected: [1, 2, 4]
console.log("V3:", linkedListToArray(merged2_3)); // Expected: [1, 2, 4]
console.log("V4:", linkedListToArray(merged2_4)); // Expected: [1, 2, 4]
console.log("V5:", linkedListToArray(merged2_5));
console.log("\n");

// Test Case 3: Both lists are empty
console.log("Test Case 3: Both Lists Empty");
let merged3_1 = mergeTwoListsV1(createLinkedList([]), createLinkedList([]));
let merged3_2 = mergeTwoListsV2(createLinkedList([]), createLinkedList([]));
let merged3_3 = mergeTwoListsV3(createLinkedList([]), createLinkedList([]));
let merged3_4 = mergeTwoListsV4(createLinkedList([]), createLinkedList([]));
let merged3_5 = mergeTwoListsV5(createLinkedList([]), createLinkedList([]));
console.log("V1:", linkedListToArray(merged3_1)); // Expected: []
console.log("V2:", linkedListToArray(merged3_2)); // Expected: []
console.log("V3:", linkedListToArray(merged3_3)); // Expected: []
console.log("V4:", linkedListToArray(merged3_4)); // Expected: []
console.log("V5:", linkedListToArray(merged3_5));
console.log("\n");

// Test Case 4: One element lists
console.log("Test Case 4: One Element Lists");
let merged4_1 = mergeTwoListsV1(createLinkedList([5]), createLinkedList([1]));
let merged4_2 = mergeTwoListsV2(createLinkedList([5]), createLinkedList([1]));
let merged4_3 = mergeTwoListsV3(createLinkedList([5]), createLinkedList([1]));
let merged4_4 = mergeTwoListsV4(createLinkedList([5]), createLinkedList([1]));
let merged4_5 = mergeTwoListsV5(createLinkedList([5]), createLinkedList([1]));
console.log("V1:", linkedListToArray(merged4_1)); // Expected: [1, 5]
console.log("V2:", linkedListToArray(merged4_2)); // Expected: [1, 5]
console.log("V3:", linkedListToArray(merged4_3)); // Expected: [1, 5]
console.log("V4:", linkedListToArray(merged4_4)); // Expected: [1, 5]
console.log("V5:", linkedListToArray(merged4_5));
console.log("\n");

// Test Case 5: Unequal length lists
console.log("Test Case 5: Unequal Length Lists");
let merged5_1 = mergeTwoListsV1(createLinkedList([1,2,3]), createLinkedList([4,5,6,7,8]));
let merged5_2 = mergeTwoListsV2(createLinkedList([1,2,3]), createLinkedList([4,5,6,7,8]));
let merged5_3 = mergeTwoListsV3(createLinkedList([1,2,3]), createLinkedList([4,5,6,7,8]));
let merged5_4 = mergeTwoListsV4(createLinkedList([1,2,3]), createLinkedList([4,5,6,7,8]));
let merged5_5 = mergeTwoListsV5(createLinkedList([1,2,3]), createLinkedList([4,5,6,7,8]));
console.log("V1:", linkedListToArray(merged5_1));
console.log("V2:", linkedListToArray(merged5_2));
console.log("V3:", linkedListToArray(merged5_3));
console.log("V4:", linkedListToArray(merged5_4));
console.log("V5:", linkedListToArray(merged5_5));
console.log("\n");

// Test case 6: one list is null
console.log("Test Case 6: one list is null");
const merged6_1 = mergeTwoListsV1(list5, null);
const merged6_2 = mergeTwoListsV2(list5, null);
const merged6_3 = mergeTwoListsV3(list5, null);
const merged6_4 = mergeTwoListsV4(list5, null);
const merged6_5 = mergeTwoListsV5(list5, null);
console.log("V1:", linkedListToArray(merged6_1));
console.log("V2:", linkedListToArray(merged6_2));
console.log("V3:", linkedListToArray(merged6_3));
console.log("V4:", linkedListToArray(merged6_4));
console.log("V5:", linkedListToArray(merged6_5));
console.log("\n");

File: C++/6-LinkedList/Problems/2-MergeTwoSortedLists/MergeTwoSortedListsPro.cpp
#include <iostream>
#include <vector>
#include <list>
#include <algorithm>
#include <memory> // For smart pointers

// Definition for singly-linked list.  Using a struct for simplicity and compatibility.
struct ListNode {
    int val;
    ListNode *next;
    // Constructor for convenience
    ListNode(int x) : val(x), next(nullptr) {}
};

// Function to create a linked list from a vector (useful for testing)
ListNode* createLinkedList(const std::vector<int>& nums) {
    ListNode* head = nullptr;
    ListNode* tail = nullptr;
    for (int num : nums) {
        ListNode* newNode = new ListNode(num);
        if (!head) {
            head = newNode;
            tail = newNode;
        } else {
            tail->next = newNode;
            tail = newNode;
        }
    }
    return head;
}

// Function to print a linked list (useful for testing)
void printLinkedList(ListNode* head) {
    ListNode* current = head;
    while (current) {
        std::cout << current->val << " -> ";
        current = current->next;
    }
    std::cout << "nullptr" << std::endl;
}

// Function to free a linked list (to prevent memory leaks)
void freeLinkedList(ListNode* head) {
    ListNode* current = head;
    while (current) {
        ListNode* next = current->next;
        delete current;
        current = next;
    }
}

// 1. Iterative Approach (Most Common, Efficient)
//   - Time Complexity: O(m+n), where m and n are the lengths of the lists.
//   - Space Complexity: O(1) - constant extra space.
//   - Real-world use case:  Merging sorted result sets from two database queries.  Very efficient and commonly used.
ListNode* mergeTwoListsIterative(ListNode* l1, ListNode* l2) {
    // Create a dummy node to simplify the logic.  This avoids special cases
    // for handling the head of the merged list.
    ListNode dummy(0);
    ListNode* tail = &dummy; // Use a pointer to the tail of the merged list.

    // Iterate while both lists have elements.
    while (l1 && l2) {
        if (l1->val <= l2->val) {
            tail->next = l1; // Add the smaller element from l1 to the merged list.
            l1 = l1->next;      // Move to the next node in l1.
        } else {
            tail->next = l2; // Add the smaller element from l2 to the merged list.
            l2 = l2->next;      // Move to the next node in l2.
        }
        tail = tail->next; // Move the tail pointer to the newly added node.
    }

    // Append the remaining elements of either list (if any).  Only one of
    // these two 'while' loops will execute.
    if (l1) {
        tail->next = l1;
    } else {
        tail->next = l2;
    }

    return dummy.next; // Return the head of the merged list (excluding the dummy node).
}

// 2. Recursive Approach
//   - Time Complexity: O(m+n)
//   - Space Complexity: O(m+n) - due to the recursive call stack.
//   - Real-world use case:  Could be used in a divide-and-conquer algorithm, though
//     the iterative approach is usually preferred for its better space efficiency.
ListNode* mergeTwoListsRecursive(ListNode* l1, ListNode* l2) {
    // Base cases: if either list is empty, return the other list.
    if (!l1) return l2;
    if (!l2) return l1;

    if (l1->val <= l2->val) {
        // Recursively merge the rest of l1 with l2.
        l1->next = mergeTwoListsRecursive(l1->next, l2);
        return l1; // Return the current node (l1) as it's the smaller.
    } else {
        // Recursively merge l1 with the rest of l2.
        l2->next = mergeTwoListsRecursive(l1, l2->next);
        return l2; // Return the current node (l2) as it's the smaller.
    }
}

// 3. Using a Vector and std::merge (STL Approach)
//   - Time Complexity: O(m+n)
//   - Space Complexity: O(m+n) - for the vector.
//   - Real-world use case:  Good for one-off merging tasks or when you need the
//     merged result in a contiguous memory block (e.g., for further processing
//     with other STL algorithms).  Not ideal if you *must* maintain a linked list.
ListNode* mergeTwoListsVector(ListNode* l1, ListNode* l2) {
    std::vector<int> mergedVector;
    // Iterate through the linked lists, adding elements to the vector.
    while (l1) {
        mergedVector.push_back(l1->val);
        l1 = l1->next;
    }
    while (l2) {
        mergedVector.push_back(l2->val);
        l2 = l2->next;
    }
    // Sort the vector.  std::sort is generally very efficient (often introsort).
    std::sort(mergedVector.begin(), mergedVector.end());

    // Alternatively, use std::merge:
    // std::vector<int> mergedVector(length of l1 + length of l2); //Pre-allocate, if size known
    // std::merge(v1.begin(), v1.end(), v2.begin(), v2.end(), mergedVector.begin());

    // Create a new linked list from the sorted vector.
    ListNode* head = nullptr;
    ListNode* tail = nullptr;
    for (int val : mergedVector) {
        ListNode* newNode = new ListNode(val);
        if (!head) {
            head = newNode;
            tail = newNode;
        } else {
            tail->next = newNode;
            tail = newNode;
        }
    }
    return head;
}

// 4. Using std::list and std::merge (STL List Approach)
//   - Time Complexity: O(m+n)
//   - Space Complexity: O(m+n) - for the lists.
//   - Real-world use case:  If you're already working with `std::list` in your project
//     (which is a doubly-linked list), this approach can be convenient.  `std::list`
//     has its own `merge` method.
ListNode* mergeTwoListsSTLList(ListNode* l1, ListNode* l2) {
    std::list<int> list1, list2, mergedList;

    // Copy the values from the input linked lists to std::list objects.
    while (l1) {
        list1.push_back(l1->val);
        l1 = l1->next;
    }
    while (l2) {
        list2.push_back(l2->val);
        l2 = l2->next;
    }
    //list1.sort(); //Ensure lists are sorted.  The input *should* be sorted, but good practice.
    //list2.sort();

     // Merge the two sorted lists using the std::list::merge() method.  This
     // method is specifically designed for merging sorted lists, and it
     // maintains the sorted order.  It's an in-place merge, meaning elements
     // are moved from list2 into list1.  After the merge, list2 will be empty.
    list1.merge(list2); // Merges list2 into list1, list2 becomes empty.
    mergedList = list1;

    // Create a new linked list from the merged std::list.
    ListNode* head = nullptr;
    ListNode* tail = nullptr;
    for (int val : mergedList) {
        ListNode* newNode = new ListNode(val);
        if (!head) {
            head = newNode;
            tail = newNode;
        } else {
            tail->next = newNode;
            tail = newNode;
        }
    }
    return head;
}

// 5. In-Place Merge (with modification of input lists, for specific scenarios)
//   - Time Complexity: O(m+n)
//   - Space Complexity: O(1) -  *if* you are allowed to modify the input lists.
//   - Real-world use case:  This is a more advanced technique.  It's useful in
//     memory-constrained environments where you cannot afford the extra space
//     for a new list.  **Important:** This modifies the original linked lists.
//     Use with caution!  This approach is similar to the iterative, but modifies the lists.
ListNode* mergeTwoListsInPlace(ListNode* l1, ListNode* l2) {
     if (!l1) return l2;
     if (!l2) return l1;

    ListNode* head = nullptr;
    ListNode* tail = nullptr;

    // Determine the head of the merged list.
    if (l1->val <= l2->val) {
        head = l1;
        tail = l1;
        l1 = l1->next;
    } else {
        head = l2;
        tail = l2;
        l2 = l2->next;
    }

    // Merge the remaining nodes.
    while (l1 && l2) {
        if (l1->val <= l2->val) {
            tail->next = l1;
            l1 = l1->next;
            tail = tail->next;
        } else {
            tail->next = l2;
            l2 = l2->next;
            tail = tail->next;
        }
    }

    // Append the remaining nodes.
    if (l1) {
        tail->next = l1;
    } else {
        tail->next = l2;
    }

    return head;
}

int main() {
    // Example usage:
    std::vector<int> nums1 = {1, 2, 4, 7, 9};
    std::vector<int> nums2 = {1, 3, 4, 6, 8, 10};

    ListNode* list1 = createLinkedList(nums1);
    ListNode* list2 = createLinkedList(nums2);

    std::cout << "List 1: ";
    printLinkedList(list1);
    std::cout << "List 2: ";
    printLinkedList(list2);
    std::cout << std::endl;

    // 1. Iterative Merge
    ListNode* mergedListIterative = mergeTwoListsIterative(list1, list2);
    std::cout << "Merged (Iterative): ";
    printLinkedList(mergedListIterative);
    freeLinkedList(mergedListIterative); // Clean up the allocated memory.
    list1 = createLinkedList(nums1);  //Recreate lists
    list2 = createLinkedList(nums2);

    // 2. Recursive Merge
    ListNode* mergedListRecursive = mergeTwoListsRecursive(list1, list2);
    std::cout << "Merged (Recursive): ";
    printLinkedList(mergedListRecursive);
    freeLinkedList(mergedListRecursive);
    list1 = createLinkedList(nums1);
    list2 = createLinkedList(nums2);

    // 3. Vector Merge
    ListNode* mergedListVector = mergeTwoListsVector(list1, list2);
    std::cout << "Merged (Vector): ";
    printLinkedList(mergedListVector);
    freeLinkedList(mergedListVector);
     list1 = createLinkedList(nums1);
    list2 = createLinkedList(nums2);

    // 4. STL List Merge
    ListNode* mergedListSTLList = mergeTwoListsSTLList(list1, list2);
    std::cout << "Merged (STL List): ";
    printLinkedList(mergedListSTLList);
    freeLinkedList(mergedListSTLList);
    list1 = createLinkedList(nums1);
    list2 = createLinkedList(nums2);

    // 5. In-Place Merge
    ListNode* mergedListInPlace = mergeTwoListsInPlace(list1, list2);
    std::cout << "Merged (In-Place): ";
    printLinkedList(mergedListInPlace);
    //The original lists are modified in-place merge.
    std::cout << "List 1 (Modified): ";
    printLinkedList(list1);
    std::cout << "List 2 (Modified): ";
    printLinkedList(list2);
    freeLinkedList(list1);  // Free, although they were modified.
    //freeLinkedList(list2); // Don't double free.  l2 is now empty after merge.
    //list1 = createLinkedList(nums1); // No need to recreate.

    return 0;
}
